core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java

```diff
@@ -0,0 +1,803 @@
+                 }
+                 return e2;
+              }, null);
+      return Optional.ofNullable(value);
+   }
+
+   @Override
+   public Optional<R> max(Comparator<? super R> comparator) {
+      R value = performOperation(TerminalFunctions.maxFunction(comparator), false,
+              (e1, e2) -> {
+                 if (e1 != null) {
+                    if (e2 != null) {
+                       return comparator.compare(e1, e2) > 0 ? e1 : e2;
+                    } else {
+                       return e1;
+                    }
+                 }
+                 return e2;
+              }, null);
+      return Optional.ofNullable(value);
+   }
+
+   @Override
+   public boolean anyMatch(Predicate<? super R> predicate) {
+      return performOperation(TerminalFunctions.anyMatchFunction(predicate), false, Boolean::logicalOr, b -> b);
+   }
+
+   @Override
+   public boolean allMatch(Predicate<? super R> predicate) {
+      return performOperation(TerminalFunctions.allMatchFunction(predicate), false, Boolean::logicalAnd, b -> !b);
+   }
+
+   @Override
+   public boolean noneMatch(Predicate<? super R> predicate) {
+      return performOperation(TerminalFunctions.noneMatchFunction(predicate), false, Boolean::logicalAnd, b -> !b);
+   }
+
+   @Override
+   public Optional<R> findFirst() {
+      if (intermediateType.shouldUseIntermediate(sorted, distinct)) {
+         Iterator<R> iterator = iterator();
+         SingleRunOperation<Optional<R>, R, Stream<R>> op = new SingleRunOperation<>(localIntermediateOperations,
+                 () -> StreamSupport.stream(Spliterators.spliteratorUnknownSize(
+                         iterator, Spliterator.CONCURRENT | Spliterator.NONNULL), parallel), s -> s.findFirst());
+         return op.performOperation();
+      } else {
+         return findAny();
+      }
+   }
+
+   @Override
+   public Optional<R> findAny() {
+      R value = performOperation(TerminalFunctions.findAnyFunction(), false, (r1, r2) -> r1 == null ? r2 : r1,
+              a -> a != null);
+      return Optional.ofNullable(value);
+   }
+
+   @Override
+   public long count() {
+      return performOperation(TerminalFunctions.countFunction(), true, (l1, l2) -> l1 + l2, null);
+   }
+
+
+   // The next ones are key tracking terminal operators
+
+   @Override
+   public Iterator<R> iterator() {
+      if (intermediateType.shouldUseIntermediate(sorted, distinct)) {
+         Stream<R> stream = performIntermediateRemoteOperation(Function.identity());
+         return stream.iterator();
+      } else {
+         return remoteIterator();
+      }
+   }
+
+   Iterator<R> remoteIterator() {
+      BlockingQueue<R> queue = new ArrayBlockingQueue<>(distributedBatchSize);
+
+      final AtomicBoolean complete = new AtomicBoolean();
+
+      Lock nextLock = new ReentrantLock();
+      Condition nextCondition = nextLock.newCondition();
+
+      Consumer<R> consumer = new HandOffConsumer(queue, complete, nextLock, nextCondition);
```

Comment:
I will double check this, also https://github.com/infinispan/infinispan/pull/3569/files#diff-be6f435a8edfd997f62d9a90ad59582bR427 would need it too.